src/System.CommandLine/ParseResult.cs

```diff
@@ -51,25 +53,7 @@ public class ParseResult
             }
 
             CommandLineText = commandLineText;
-
-            if (unmatchedTokens is null)
-            {
-                _unmatchedTokens = Array.Empty<Token>();
-            }
-            else
-            {
-                _unmatchedTokens = unmatchedTokens;
-
-                if (parser.Configuration.RootCommand.TreatUnmatchedTokensAsErrors)
```

Comment:
we now do it as soon as the unmatched token is found. Bonus: better ordering of errors